draft-irtf-cfrg-opaque.md

```diff
@@ -1385,22 +1385,68 @@ following fields:
-  deterministically from the input `seed`. This function is implemented as
-  DeriveKeyPair(seed, "OPAQUE-DeriveAuthKeyPair"), where DeriveKeyPair is
-  as specified in {{OPRF, Section 3.2}}.
+  deterministically from the input `seed`.
 - GenerateAuthKeyPair(): Return a randomly generated private and public key
   pair. This can be implemented by invoking DeriveAuthKeyPair with `Nseed`
   random bytes as input.
-- SerializeElement(element): A member function of the underlying group that
-  maps `element` to a unique byte array, mirrored from the definition of the
-  similarly-named function of the OPRF group described in
-  {{OPRF, Section 2.1}}.
+- SerializeElement(element): A function that maps `element` to a unique byte
+  array, mirrored from the definition of the similarly-named function of the
+  OPRF group described in {{OPRF, Section 2.1}}.
+- ScalarMult(k, B): A function that performs scalar multiplication between the
+  scalar input `k` and element `B`.
+
+Implementations for recommended groups in {{configurations}}, as well as groups
+covered by test vectors in {{test-vectors}}, are described in the following sections.
+
+#### ristretto255 Group
+
+This section describes the implementation of the AKE group based on ristretto255,
+as defined in {{?RISTRETTO=I-D.irtf-cfrg-ristretto255-decaf448}}.
+
+- DeriveAuthKeyPair(seed): This function is implemented as
+  DeriveKeyPair(seed, "OPAQUE-DeriveAuthKeyPair"), where DeriveKeyPair is
+  as specified in {{OPRF, Section 3.2}}.
+- GenerateAuthKeyPair(): This is implemented by invoking DeriveAuthKeyPair
+  with `Nseed` random bytes as input.
+- SerializeElement(element): This is implemented using the similarly-named
+  function of the OPRF group described in {{OPRF, Section 2.1}}.
+- ScalarMult(k, B): Implemented as scalar multiplication as described in
+  {{Section 4 of RISTRETTO}}.
+
+#### P-256 Group
+
+This section describes the implementation of the AKE group based on NIST P-256,
+as defined in {{?NISTCurves=DOI.10.6028/NIST.FIPS.186-4}}.
+
+- DeriveAuthKeyPair(seed): This function is implemented as
+  DeriveKeyPair(seed, "OPAQUE-DeriveAuthKeyPair"), where DeriveKeyPair is
+  as specified in {{OPRF, Section 3.2}}.
+- GenerateAuthKeyPair(): This is implemented by invoking DeriveAuthKeyPair
+  with `Nseed` random bytes as input.
+- SerializeElement(element): This is implemented using the similarly-named
+  function of the OPRF group described in {{OPRF, Section 2.1}}.
+- ScalarMult(k, B): Implemented as scalar multiplication as described in
+  {{NISTCurves}}.
+
+#### Curve25519 Group
+
+This section describes the implementation of the AKE group based on Curve25519,
+as defined in {{?Curve25519=RFC7748}}.
+
+- DeriveAuthKeyPair(seed): This function is implemented by returning the private
+  key `k` based on seed (of length `Nseed = 32` bytes), as described in {{Section 5 of Curve25519}},
+  as well as the result of ScalarMult(k, B), where B is the base point of Curve25519.
```

Comment:
Why does this differ from the other groups?
Because OPRF doesn't specify a ciphersuite for Curve25519?

Reply:
Yeah, exactly.

Reply:
After consulting the spec again, I couldn't really find an answer to this. Doesn't removing the `G.HashToScalar` have any security implications at all?
And if not, why is it in place for the other ciphersuites?

Reply:
I'm not sure I understand the question. But to try and reiterate, there was no existing function in the OPRF document we could point to, which is why we just lean on standard key generation for curve25519 -- treat random bytes as the key (scalar).

Reply:
Ah, sorry, I can see how my question is confusing. So in `opaque-ke` we were using OPRF with a custom ciphersuite when chossing Curve25519, which was possible because the hash2curve specification does provide a ciphersuite for Curve25519.

So my question is: the other ciphersuites do use OPRF to derive a key pair in `DeriveAuthKeyPair`. So why do other ciphersuites need that but Curve25519 doesn't? I thought it was necessary to for security reasons (that I have no clue about).

Reply:
Oh, I see! You're asking if it's necessary to map the input seed to a scalar using something like a random oracle? I don't _think_ it's necessary, but I actually don't know the answer off the top of my head. I think this should probably be asked on the list (I can do that).

Reply:
Thank you!